src/gui/src/layoutViewer.cpp

```diff
@@ -205,53 +203,173 @@ class GuiPainter : public Painter
+      sy += text_bbox.height() / 2;
+    } else {
+      // RIGHT_CENTER
+      sx -= text_bbox.right();
+      sy += text_bbox.height() / 2;
+    }
+    // current units of sx, sy are pixels, so convert to DBU
+    sx *= scale_adjust;
+    sy *= scale_adjust;
+    // add desired text location in DBU
+    sx += x;
+    sy += y;
+    const QTransform transform = painter_->transform();
+    painter_->translate(sx, sy);
+    painter_->scale(scale_adjust, -scale_adjust); // undo original scaling
+    painter_->drawText(0, 0, text); // origin of painter is desired location, so paint at 0, 0
+    painter_->setTransform(transform);
+  }
+
+  void drawRuler(int x0, int y0, int x1, int y1, const std::string& label = "") override
   {
+    const QColor ruler_color_qt = getOptions()->rulerColor();
+    const Color ruler_color(ruler_color_qt.red(), ruler_color_qt.green(), ruler_color_qt.blue(), ruler_color_qt.alpha());
+    const QFont ruler_font = getOptions()->rulerFont();
+    const QFont restore_font = painter_->font();
+
     setPen(ruler_color, true);
     setBrush(ruler_color);
 
     std::stringstream ss;
-    const int x_len = std::abs(x0 - x1);
-    const int y_len = std::abs(y0 - y1);
+    const double x_len = x1 - x0;
+    const double y_len = y1 - y0;
+    const double len = std::sqrt(x_len * x_len + y_len * y_len);
+    if (len == 0) {
+      // zero length ruler
+      return;
+    }
+
+    const QTransform initial_xfm = painter_->transform();
+
+    painter_->translate(x0, y0);
+    qreal ruler_angle;
+    if (x_len == 0) {
+      if (y1 > y0) {
+        ruler_angle = 90;
+      } else {
+        ruler_angle = -90;
+      }
+    } else if (y_len == 0) {
+      if (x1 > x0) {
+        ruler_angle = 0;
+      } else {
+        ruler_angle = -180;
+      }
+    } else {
+      ruler_angle = 57.295779 * std::atan(std::abs(y_len / x_len)); // 180 / pi
+      if (x_len < 0) { // adjust for negative dx
+        ruler_angle = 180 - ruler_angle;
+      }
+      if (y_len < 0) { // adjust for negative dy
+       ruler_angle = -ruler_angle;
+      }
+    }
+    painter_->rotate(ruler_angle);
 
     const int precision = std::ceil(std::log10(dbu_per_micron_));
+    const qreal len_microns = len / (qreal) dbu_per_micron_;
 
-    ss << std::fixed << std::setprecision(precision)
-       << std::max(x_len, y_len) / (qreal) dbu_per_micron_;
+    ss << std::fixed << std::setprecision(precision) << len_microns;
 
-    drawLine(x0, y0, x1, y1);
+    const bool flip_direction = -90 >= ruler_angle || ruler_angle > 90;
 
-    if (x_len < y_len) {
-      drawString(x0, (y0 + y1) / 2, 0, ss.str());
+    // draw center line
+    drawLine(0, 0, len, 0);
+    // draw endcaps (arrows) (5 px or 2 DBU if very close)
+    int endcap_size = std::max(2.0, 5.0 / getPixelsPerDBU());
+    if (flip_direction) {
+      endcap_size = -endcap_size;
+    }
+    drawLine(0, -endcap_size, 0, 0);
+    drawLine(len, -endcap_size, len, 0);
+
+    // tick mark interval in microns
```

Comment:
Can the ruler & the scale bar share code?

Reply:
I suppose it would be possible. Main difference is that the rulers are drawn in DBU and scale bar in pixels (so should be doable) is there one you prefer over the other?

Reply:
I haven't looked closely enough to see if combining, while handling dbu and pixels, would make things simpler or more complex .  I'll leave it to your best judgement.